### PR TITLE
ASoC: Intel: sof_sdw: starts non sdw BE id with the largest sdw BE id

### PR DESCRIPTION
**Pull request by bardliao**
We could create sdw links with larger BE id first, in that case non sdw links BE id will start with the latest sdw BE id instead of the largest sd BE id. As a result, some BE ids may be duplicated.

For example,
create dai link SDW0-Playback-SimpleJack, id 0
create dai link SDW0-Capture-SmartMic, id 4
create dai link SDW0-Capture-SimpleJack, id 1
create dai link SDW2-Playback-SmartAmp, id 2
create dai link SDW2-Capture-SmartAmp, id 3
create dai link iDisp1, id 4
create dai link iDisp2, id 5
create dai link iDisp3, id 6

### CODE REVIEW

**Review by plbossart — DISMISSED** [review]

**Review by aiChaoSONG — APPROVED** [review]
On the device the issue is reproduced, issue can be fixed with this patch.
On the device with rt711+rt1316+rt714 codecs, this patch has no impact.

File: sound/soc/intel/boards/sof_sdw.c [hunk 1]
```diff
@@ -1786,13 +1788,20 @@ static int sof_card_dai_links_create(struct snd_soc_card *card)
 					dev_err(dev, "failed to create dai link %d\n", link_index);
 					return ret;
 				}
+				if (be_id > largest_sdw_be_id)
+					largest_sdw_be_id = be_id;
```
Comment (ujfalusi) [inline sound/soc/intel/boards/sof_sdw.c]: Why not introduce a variable under this for loop and update the be_id, something like that:
```diff
diff --git a/sound/soc/intel/boards/sof_sdw.c b/sound/soc/intel/boards/sof_sdw.c
index ef741907b64d..6be54270e7fd 100644
--- a/sound/soc/intel/boards/sof_sdw.c
+++ b/sound/soc/intel/boards/sof_sdw.c
@@ -1777,15 +1777,24 @@ static int sof_card_dai_links_create(struct snd_soc_card *card)
 				return codec_index;
 
 			for (j = 0; j < codec_info_list[codec_index].dai_num ; j++) {
+				int current_be_id;
+
 				ret = create_sdw_dailink(card, &link_index, dai_links,
 							 sdw_be_num, adr_link,
 							 codec_conf, codec_conf_num,
-							 &be_id, &codec_conf_index,
+							 &current_be_id, &codec_conf_index,
 							 &ignore_pch_dmic, append_dai_type, i, j);
 				if (ret < 0) {
 					dev_err(dev, "failed to create dai link %d\n", link_index);
 					return ret;
 				}
+
+				/*
+				 * Update the be_id to match the highest ID used
+				 * for SDW link
+				 */
+				if (be_id < current_be_id)
+					be_id = current_be_id;
 			}
 
 			if (aggregation && endpoint->aggregated)
@@ -1793,6 +1802,8 @@ static int sof_card_dai_links_create(struct snd_soc_card *card)
 		}
 	}
 
+	/* Increment the be_id for non SDW links to avoid overlap */
+	be_id++;
 SSP:
 	/* SSP */
 	if (!ssp_num)
```

File: sound/soc/intel/boards/sof_sdw.c [hunk 2]
```diff
@@ -1786,13 +1788,20 @@ static int sof_card_dai_links_create(struct snd_soc_card *card)
...
 				group_generated[endpoint->group_id] = true;
 		}
 	}
-
+	/*
+	 * We could create sdw links with larger BE id first, in that case non sdw links be id
+	 * will start with the latest BE id instead of the largest BE id. As a result, some
+	 * BE ids may be duplicated.
+	 */
+	be_id = largest_sdw_be_id;
```
Comment (ujfalusi) [inline sound/soc/intel/boards/sof_sdw.c]: Should be `be_id = largest_sdw_be_id + 1;`?
But see my diff which takes care of this

File: sound/soc/intel/boards/sof_sdw.c [hunk 3]
```diff
@@ -1786,13 +1788,20 @@ static int sof_card_dai_links_create(struct snd_soc_card *card)
...
 
 			if (aggregation && endpoint->aggregated)
 				group_generated[endpoint->group_id] = true;
 		}
 	}
-
+	/*
+	 * We could create sdw links with larger BE id first, in that case non sdw links be id
+	 * will start with the latest BE id instead of the largest BE id. As a result, some
+	 * BE ids may be duplicated.
```
Comment (ujfalusi) [inline sound/soc/intel/boards/sof_sdw.c]: I find this comment hard to follow, similarly to the commit message.
The issue is: The soundwire links do not have their IDs as consecutive numbers, thus the last link might have lower be_id than the previous one and this leads to id collision with non SDW links.

**Review by ujfalusi — CHANGES_REQUESTED** [review]

File: sound/soc/intel/boards/sof_sdw.c (hunk 2 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > Should be `be_id = largest_sdw_be_id + 1;`? But see my diff which takes care of this

No, be_id is increased in init_dai_link(). So, `current_be_id` is already ++.

**Comment by bardliao** [comment]
Thanks @ujfalusi, updated.

**Review by plbossart — APPROVED** [review]

**Review by aiChaoSONG — APPROVED** [review]

**Review by ujfalusi — APPROVED** [review]
